### PR TITLE
[QC-262] Data Sampling Benchmark

### PR DESCRIPTION
**Pull request by knopers8**
This introduces a script which runs Data Sampling benchmarks and prints results to files. It is supposed to work both with zeromq and shmem transport (as enabled in this PR: https://github.com/AliceO2Group/AliceO2/pull/2670 ).
There is a mechanism which prevents the benchmark from allocating too much memory, to avoid deadlocks (Dispatcher need to copy data, but there is not enough memory for that and it waits until there is, but only Dispatcher can free it by processing messages).
However, with that fixed, I can see still some crashes related to shared memory (appearing only for larger throughputs/message payload sizes):
```
[193856:Dispatcher]:  *** Break *** bus error
[193856:Dispatcher]: essor_id=Dispatcher
...
[193856:Dispatcher]: ===========================================================
[193856:Dispatcher]: 
[193856:Dispatcher]: 
[193856:Dispatcher]: The lines below might hint at the cause of the crash.
[193856:Dispatcher]: You may get help by asking at the ROOT forum http://root.cern.ch/forum
[193856:Dispatcher]: Only if you are really convinced it is a bug in ROOT then please submit a
[193856:Dispatcher]: report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
[193856:Dispatcher]: from above as an attachment in addition to anything else
[193856:Dispatcher]: that might help us fixing this issue.
[193856:Dispatcher]: ===========================================================
[193856:Dispatcher]: #7  0x00007fc9308a9b5a in __memcpy_ssse3_back () from /lib64/libc.so.6
[193856:Dispatcher]: #8  0x00007fc93bd25549 in o2::framework::DataAllocator::snapshot (this=0x20e2de0, spec=..., payload=0x7fb3d997c350 "", payloadSize=100000000, serializationMethod=...) at /home/datasam
pling/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/DataAllocator.cxx:186
[193856:Dispatcher]: #9  0x00007fc93bd653ce in o2::framework::Dispatcher::send (this=this
[193856:Dispatcher]: entry=0x2104bc0, dataAllocator=..., inputData=..., output=...) at /home/datasampling/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/Dispatcher.cxx:157
[193856:Dispatcher]: #10 0x00007fc93bd67731 in o2::framework::Dispatcher::run (this=0x2104bc0, ctx=...) at /home/datasampling/alice/sw/SOURCES/O2/0_O2_DATAFLOW/0/Framework/Core/src/Dispatcher.cxx:99
```
@dennisklein @rbx Have you seen perhaps something like that before? Could it happen because I open a shmem segment to monitor the free memory used by my workflow?  (see dataSamplingBenchmark.cxx:227)

### CODE REVIEW

**Comment by rbx** [comment]
I have not encountered the bus error like that before.

I don't think that opening the segment in that way leads to this problem (although copying our private methods is not guaranteed to work in the future). Even if you call your checker after the segment is already closed by the transport (after ResetDevice state), it should give a correct response (just tested it to make sure). If you call it before the segment is created, you should get an "no such file or directory" exception.

**Comment by knopers8** [comment]
> (although copying our private methods is not guaranteed to work in the future)

Yes, I am not satisfied with that neither. I would like to replace it with a more maintainable solution, but I didn't find out yet how to get this ID more easily.

> If you call it before the segment is created, you should get an "no such file or directory" exception.

Indeed, this is what I would get, this is why I put a sleep() before creating a segment on the start (which also isn't ideal).

Then, would you perhaps have any pointers where one should start look for the cause? Might it be something in FairMQ or boost and how it manages the memory?

**Comment by rbx** [comment]
Do you perhaps have a more detailed log/backtrace? Showing which FairMQ/boost call could have cause the call. It only happens when you try to monitor free use? Does it happen in the same process where you have the usage monitor, or in any other shmem users? I found an [ancient post](https://stackoverflow.com/questions/3103255/boostinterprocess-shared-memory-bus-error?rq=1) related to boost::interprocess and `bus error`, perhaps it contains some hint.

**Comment by knopers8** [comment]
Thank you for the help, that ancient post was actually helpful! I wasn't aware that /dev/shm gets a half of total RAM (63GB out of 128GB in my case), while I specified a shmem segment size of 100GB. The bus error was occurring just when the available memory on /dev/shm was on the limit:
```
Filesystem      Size  Used Avail Use% Mounted on                                                     │    11 root      rt   0       0      0      0 S   0.0  0.0   0:06.63 [watchdog/0]                     
tmpfs            63G   61G  2.0G  97% /dev/shm                                                       ├──────────────────────────────────────────────────────────────────────────────────────────────────────
Filesystem      Size  Used Avail Use% Mounted on                                                     │WritebackTmp:          0 kB                                                                   [3/1776]
tmpfs            63G   62G  1.2G  99% /dev/shm                                                       │CommitLimit:    73999484 kB
Filesystem      Size  Used Avail Use% Mounted on                                                     │Committed_AS:     449752 kB
tmpfs            63G   63G  357M 100% /dev/shm                                                       │VmallocTotal:   34359738367 kB
Filesystem      Size  Used Avail Use% Mounted on                                                     │VmallocUsed:      542544 kB
tmpfs            63G   63G     0 100% /dev/shm     
```
Reducing the shm segment size helped. But I guess we will have to modify that limit in production.

There were no related calls in FairMQ/Boost, just that memcpy, which actually makes sense.

**Comment by knopers8** [comment]
@Barthelemy This is ready for review. The test for multiple Dispatchers will fail until https://github.com/AliceO2Group/AliceO2/pull/2950/ draws to an end. These benchmarks are expected to run on servers, it might fail on laptops with 1-2 GBs RAM left available.

File: Framework/Core/scripts/o2-datasampling-benchmark.sh [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+# Run the benchmark with given parameters
+# \param 1 : fractions array name
+# \param 2 : payload sizes array name
+# \param 3 : number of producers array name
+# \param 4 : number of dispatchers array name
+# \param 5 : repetitions
+# \param 6 : test duration
+# \param 7 : warm up cycles - how many first metrics should be ignored (they are sent each 10s)
+# \param 8 : test name
+# \param 9 : free_memory_bytes
```
Comment (Barthelemy) [inline Framework/Core/scripts/o2-datasampling-benchmark.sh]: what is free_memory_bytes ?

Comment (knopers8) [reply Framework/Core/scripts/o2-datasampling-benchmark.sh]: added explanation

**Comment by knopers8** [comment]
@Barthelemy All green except for Codacy, which couldn't detected that two variables *are* actually used.

**Pull request merged**